src/frontend/src/binder/expr/column.rs

```diff
@@ -59,6 +62,28 @@ impl Binder {
                     err = e;
                 }
             }
+            for LateralBindContext {
+                context,
+                is_visible,
+            } in lateral_contexts
+            {
+                if *is_visible {
+                    match context.get_column_binding_index(table_name, column_name) {
+                        Ok(index) => {
+                            let column = &context.columns[index];
+                            return Ok(CorrelatedInputRef::new(
+                                column.index,
+                                column.field.data_type.clone(),
+                                depth,
+                            )
+                            .into());
+                        }
+                        Err(e) => {
+                            err = e;
+                        }
+                    }
+                }
+            }
```

Comment:
Seems this change can be deferred as part of #3849 ?

In addition, lateral context may (unconfirmed) have higher priority than upper context, and lateral context of upper context may also need to be checked.

Reply:
Indeed. we may instead want to collect all references from all contexts (upper and lateral) and throw an "ambiguous reference" error?

Reply:
Maybe we can just mark the lateral code path as dead and keep this code so it'll be easier to pick up in the future when we decide to support lateral. I'll add comments and reference to the issue/PR when marking the code path as dead.

Reply:
Added the tracking issue to the error.

Reply:
Let's keep the update to this part minimal, as the dead code is not well tested and may become outdated. For example, there should be no ambiguity for column names at different levels: the innermost one wins.

Reply:
And what if there are two levels of subquery? Would the middle one still be unambiguous?

Reply:
According to postgres, yes:
```sql
explain select * from t1 where x > (select sum(x) from t2 where z = (select y from t3 where x = y));
                                QUERY PLAN                                
--------------------------------------------------------------------------
 Seq Scan on t1  (cost=94675.79..94717.66 rows=850 width=4)
   Filter: (x > $1)
   InitPlan 2 (returns $1)
     ->  Aggregate  (cost=94675.78..94675.79 rows=1 width=8)
           ->  Seq Scan on t2  (cost=0.00..94675.75 rows=11 width=4)
                 Filter: (z = (SubPlan 1))
                 SubPlan 1
                   ->  Seq Scan on t3  (cost=0.00..41.88 rows=13 width=4)
                         Filter: (t2.x = y)
```

Reply:
Behaviour is somewhat inconsistent in FROM subquery:
```sql
explain select * from t1 where x > (select sum(x) from t2 join (select y from t3 where x = y) as s on true);
                                  QUERY PLAN                                   
-------------------------------------------------------------------------------
 Seq Scan on t1  (cost=0.00..1313846.50 rows=850 width=4)
   Filter: (x > (SubPlan 1))
   SubPlan 1
     ->  Aggregate  (cost=515.21..515.22 rows=1 width=8)
           ->  Nested Loop  (cost=0.00..441.76 rows=29380 width=4)
                 ->  Seq Scan on t2  (cost=0.00..32.60 rows=2260 width=4)
                 ->  Materialize  (cost=0.00..41.94 rows=13 width=0)
                       ->  Seq Scan on t3  (cost=0.00..41.88 rows=13 width=0)
                             Filter: (t1.x = y)
```

Reply:
What about this?
* Try not to make this part more complicated, if we do not need to support that complicated feature (lateral) yet.
* If you believe the main branch implementation contains a bug, maybe we can fix it separately, with the failing test case added.

Reply:
I've removed the changes to this area entirely.

Reply:
Yes, we will not deal with lateral.

There is no bug here, just comfirming the binding behaviour for postgres, which agrees with this statement:

> In addition, lateral context may (unconfirmed) have higher priority than upper context, and lateral context of upper context may also need to be checked.

To be implemented in future when we need it